src/main/java/com/google/cloud/spanner/jdbc/JdbcResultSet.java

```diff
@@ -115,61 +115,219 @@ private boolean isNull(int columnIndex) {
     return wasNull;
   }
 
-  private boolean isNull(String columnName) {
-    wasNull = spanner.isNull(columnName);
-    return wasNull;
+  SQLException createInvalidToGetAs(String sqlType, Code type) {
+    return JdbcSqlExceptionFactory.of(
+        String.format("Invalid column type to get as %s: %s", sqlType, type.name()),
+        com.google.rpc.Code.INVALID_ARGUMENT);
   }
 
   @Override
   public String getString(int columnIndex) throws SQLException {
     checkClosedAndValidRow();
-    return isNull(columnIndex) ? null : spanner.getString(columnIndex - 1);
+    boolean isNull = isNull(columnIndex);
+    int spannerIndex = columnIndex - 1;
+    Code type = spanner.getColumnType(spannerIndex).getCode();
+    switch (type) {
+      case BOOL:
+        return isNull ? null : String.valueOf(spanner.getBoolean(spannerIndex));
+      case BYTES:
+        return isNull ? null : spanner.getBytes(spannerIndex).toBase64();
+      case DATE:
+        return isNull ? null : spanner.getDate(spannerIndex).toString();
+      case FLOAT64:
+        return isNull ? null : Double.toString(spanner.getDouble(spannerIndex));
+      case INT64:
+        return isNull ? null : Long.toString(spanner.getLong(spannerIndex));
+      case STRING:
+        return isNull ? null : spanner.getString(spannerIndex);
+      case TIMESTAMP:
+        return isNull ? null : spanner.getTimestamp(spannerIndex).toString();
+      case STRUCT:
+      case ARRAY:
+      default:
+        throw createInvalidToGetAs("string", type);
+    }
   }
 
   @Override
   public boolean getBoolean(int columnIndex) throws SQLException {
     checkClosedAndValidRow();
-    return isNull(columnIndex) ? false : spanner.getBoolean(columnIndex - 1);
+    boolean isNull = isNull(columnIndex);
+    int spannerIndex = columnIndex - 1;
+    Code type = spanner.getColumnType(spannerIndex).getCode();
+    switch (type) {
+      case BOOL:
+        return isNull ? false : spanner.getBoolean(spannerIndex);
+      case FLOAT64:
+        return isNull ? false : spanner.getDouble(spannerIndex) != 0D;
+      case INT64:
+        return isNull ? false : spanner.getLong(spannerIndex) != 0L;
+      case STRING:
+        return isNull ? false : Boolean.valueOf(spanner.getString(spannerIndex));
+      case BYTES:
+      case DATE:
+      case STRUCT:
+      case TIMESTAMP:
+      case ARRAY:
+      default:
+        throw createInvalidToGetAs("boolean", type);
+    }
   }
 
   @Override
   public byte getByte(int columnIndex) throws SQLException {
     checkClosedAndValidRow();
-    long val = isNull(columnIndex) ? 0L : spanner.getLong(columnIndex - 1);
-    return checkedCastToByte(val);
+    boolean isNull = isNull(columnIndex);
+    int spannerIndex = columnIndex - 1;
+    Code type = spanner.getColumnType(spannerIndex).getCode();
+    switch (type) {
+      case BOOL:
+        return isNull ? (byte) 0 : spanner.getBoolean(spannerIndex) ? (byte) 1 : 0;
```

Comment:
nit: Here and below, the precedence rules probably does the right thing but I think adding brackets for scoping the conditionals will help with readability. Feel free to ignore if this is the convention in Java.

Reply:
Done.